src/hermes/commands/harvest/git.py

```diff
@@ -0,0 +1,84 @@
+import glob
+import os
+import json
+import pathlib
+import urllib.request
+import typing as t
+
+import jsonschema
+import click
+import subprocess
+import shutil
+
+from hermes.model.context import HermesHarvestContext
+from hermes.model.errors import HermesValidationError
+
+# TODO: can and should we get this somehow?
+SHELL_ENCODING = 'utf-8'
+
+class AuthorData:
+    def __init__(self, line):
+        self.name = line[0]
+        self.email = set((line[1],))
+        self.tFirst = line[2]
+        self.tLast = self.tFirst
+
+    def update(self, line: t.List):
+        assert(self.name == line[0])
+
+        self.email.add(line[1])
+        t = line[2]
+        if t < self.tFirst:
+            self.tFirst = t
+        elif t > self.tFirst:
+            self.tLast = t
+
+def harvest_git(click_ctx: click.Context, ctx: HermesHarvestContext):
+    """
+    Implementation of a harvester that provides autor data from Git.
+
+    :param click_ctx: Click context that this command was run inside (might be used to extract command line arguments).
+    :param ctx: The harvesting context that should contain the provided metadata.
+    """
+    # Get the parent context (every subcommand has its own context with the main click context as parent)
+    parent_ctx = click_ctx.parent
+    if parent_ctx is None:
+        raise RuntimeError('No parent context!')
+    path = parent_ctx.params['path']
+
+    gitExe = shutil.which('git')
+    if not gitExe:
+        raise RuntimeError('Git not available!')
+
+    p = subprocess.run([gitExe, "rev-parse", "--abbrev-ref", "HEAD"], capture_output=True)
+    if p.returncode:
+        raise RuntimeError("`git branch` command failed with code {}: '{}'!".format(p.returncode, p.stderr.decode(SHELL_ENCODING)))
+    gitBranch = p.stdout.decode(SHELL_ENCODING).strip()
+    # TODO: should we warn or error if the HEAD is detached?
+
+    # Get history of currently checked-out branch
+    authors = {}
+    p = subprocess.run([gitExe, "log", "--pretty=%an_%ae_%ad", "--date=unix"], capture_output=True)
+    if p.returncode:
+        raise RuntimeError("`git log` command failed with code {}: '{}'!".format(p.returncode, p.stderr.decode(SHELL_ENCODING)))
+
+    log = p.stdout.decode(SHELL_ENCODING).split('\n')
+    for l in log:
+        d = l.split('_')
+        if len(d) != 3:
+            continue
+        try:
+            d[2] = int(d[2])
+        except ValueError:
+            continue
+
+        if d[0] in authors:
+            authors[d[0]].update(d)
+        else:
+            authors[d[0]] = AuthorData(d)
+    
+    for a in authors.values():
+        ctx.update("author.since", a.tFirst, name=a.name, branch=gitBranch)
+        ctx.update("author.until", a.tLast, name=a.name, branch=gitBranch)
+        for e in a.email:
+            ctx.update("author.email", e, name=a.name, branch=gitBranch, email=e)
```

Comment:
Based on software citation principles, the semantics are off for updating values with the `author` prefix. This is because the only thing we really know about these entities (could include bots) is that they have pushed commits. Authors can only be defined by the software project itself, so we cannot rely on git here. Perhaps update the context with the prefix `committer` instead? 

It would probably also be more "correct" to rename the vars.

Reply:
git has separate concepts for "author" and "committer". So, if someone commits someone else's code, and use git properly, the info will be correct. Committers could be harvested additionally.

Reply:
Incidentally, see https://github.com/hermes-hmc/workflow/pull/25/commits/acfdf826c27d6d550ee4bb8f92dcf8ffd489afd0 .

Reply:
> git has separate concepts for "author" and "committer". So, if someone commits someone else's code, and use git properly, the info will be correct. Committers could be harvested additionally.

Ah, I see. In this case, we should find some other term to clarify that these are *authoritative authors* as defined by the humans in the project, or what git thinks is an author. Perhaps "contributor" is better, as we definitely have contributions here.

Reply:
contributor?

Reply:
The harvested data in the cache is pre-curation anyway. There must be a curation step in which the user decides how to use the harvested data. Consider also inconsistent spelling of names or multiple listed email addresses. For the cache, it may be better stick to the terminology used in the data source to avoid confusion.

Reply:
Okay, this seems like a good argument ... even though this could be done automagically by the associated "Pre-Processor" and not during manual curation.  But this needs to be implemented yet and I think it's a beautiful example to investigate into implementation strategies for this.
So (except for the PEP-8-flaws) I like your contribution and am happy to approve it :)